src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,749 @@
+            self.amplitude_contrast, np.sqrt(1 - np.square(self.amplitude_contrast))
+        )
+        chi = (
+            defocus_factor
+            - np.pi
+            * np.power(self.lmbd, 3)
+            * self.cs
+            * np.power(10, 6)
+            * np.square(self.r_ctf)
+            / 2
+            + amplitude_contrast_term
+        )
+        h = -np.sin(chi)
+        self.h = h
+
+    def ctf_preprocess(self, micrograph, block_size):
+        """
+        Preprocess CTF of micrograph using block_size.
+
+        :param micrograph: Micrograph as numpy array. #NOTE looks like F order
+        :param blocksize: Size of the square blocks to partition micrograph.
+        :return: Numpy array of blocks.
+        """
+
+        # verify block_size is even
+        block_size = block_size - (block_size % 2)
+
+        size_x = micrograph.shape[1]
+        size_y = micrograph.shape[0]
+
+        step_size = block_size // 2
+        range_y = size_y // step_size - 1
+        range_x = size_x // step_size - 1
+
+        block_list = [
+            micrograph[
+                i * step_size : (i + 2) * step_size, j * step_size : (j + 2) * step_size
+            ]
+            for i in range(range_y)
+            for j in range(range_x)
+        ]
+        block = np.asarray(block_list, micrograph.dtype)
+
+        block_sum = np.sum(np.sum(block, axis=-1), axis=-1)
+        block_sum = np.reshape(block_sum, (block.shape[0], 1, 1))
+        block_sum = np.tile(block_sum, (1, block_size, block_size))
+        block = block - (
+            block_sum / (512 ** 2)
+        )  # equals to the matlab version (11-7-19)
+
+        return block
+
+    def ctf_tapers(self, N, R, L):
+        """
+
+        :param N:
+        :param R:
+        :param L:
+        :return:
+        """
+
+        k, el = np.meshgrid(
+            np.arange(N, dtype=self.dtype), np.arange(N, dtype=self.dtype)
+        )
+
+        denom = np.pi * (k - el)
+        denom = denom + np.eye(N, dtype=self.dtype)
+        phi_R = np.divide(np.sin(np.pi * R * (k - el)), denom)
+        np.fill_diagonal(phi_R, 1)  # absolute difference from Matlab 10^-18
+
+        data_tapers, _ = eigs(phi_R, L)
+
+        return data_tapers
+
+    def ctf_estimate_psd(self, blocks, tapers_1d, num_1d_tapers):
+        """
+
+        :param blocks:
+        :param tapers_1d:
+        :param num_1d_tapers:
+        :return: Numpy array.
+        """
+
+        tapers_1d = tapers_1d.astype(complex_type(self.dtype), copy=False)
+
+        blocks_mt_pre_fft = np.zeros(
+            blocks[0, :, :].shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt_post_fft = np.zeros(
+            blocks_mt_pre_fft.shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt = np.zeros(blocks_mt_pre_fft.shape, dtype=self.dtype)
+
+        blocks_tapered = np.zeros(blocks[0, :, :].shape, dtype=complex_type(self.dtype))
+
+        taper_2d = np.zeros(
```

Comment:
@janden , `blocks_tapers` and `taper_2d` were both uninitialized (np.empty, potentially junk).  `blocks_tapered` should be densely overwritten by the np.multiply on L239; sure.  However, the right multiplicand in L239 is `tapers_2d`, which is also potentially junk; no good. This is immediately fed that into the fft and used in the remainder of the program.  When I zero'd these, the signal arrays in the remainder of the program were much more stable run to run.

Reply:
But just above it (L233), `taper_2d` gets filled out by the `np.matmul` call, no? Maybe only the real part gets written to. Just trying to make sense of this.

Reply:
Yes, I see what you mean.  Let me check that array.

Reply:
Hrrm, maybe this was not it. I checked the shapes very carefully and I believe you are correct. If I run it enough times I can still occasionally get noise here.  What I don't understand is why changing those lines changes the frequency with which I get noise. Ugh.

Reply:
The only hot spot before that is `eigs` which I know is not deterministic, but usually doesn't cause us this much trouble.